### PR TITLE
Sycl kernel api

### LINKED ISSUE #943: dpctl doesn't support the cl::sycl::kernel in pybind11
I met issue when integrate the cl::sycl::kernel into python.

The error log like this:
```
TypeError: Unable to convert function return value to a Python type! The signature was
	(arg0: str, arg1: str, arg2: int, arg3: dpctl.SyclDevice) -> Tuple[int, cl::sycl::kernel, int, int]
```
It seems the sycl::device has been integrated successfully. But the cl::sycl::kernel is not supported.

Is there any plan to support the cl::sycl::kernel?

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
This PR add Cython generated c-api for `dpctl.program.SyclKernel` and `dpctl.program.SyclProgram` classes.

dpctl4pybind11 adds type casters to map `dpctl.program.SyclKernel` to `sycl::kernel` and 
`dpctl.program.SyclProgram` to `sycl::kernel_bundle<sycl::bundle_state::executable>`. 

This PR closes #943.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
@ndgrigorian The `DPCTLKernel_GetName` used to exist before, see https://github.com/IntelPython/dpctl/blob/0.12.0/libsyclinterface/include/dpctl_sycl_kernel_interface.h#L41-L52

I removed while refactoring from use of deprecated `sycl::program` to SYCL-2020 compliant `sycl::kernel_bundle` in #845 which made it into 0.13.0

File: dpctl/apis/include/dpctl4pybind11.hpp [hunk 1]
```diff
@@ -506,6 +532,86 @@ template <> struct type_caster<sycl::event>
...
+            value = std::make_unique<
+                sycl::kernel_bundle<sycl::bundle_state::executable>>(
+                *(reinterpret_cast<
+                    sycl::kernel_bundle<sycl::bundle_state::executable> *>(
+                    KBRef)));
+            return true;
+        }
+        else {
+            throw py::type_error(
+                "Input is of unexpected type, expected dpctl.SyclEvent");
```
Comment (ndgrigorian) [inline dpctl/apis/include/dpctl4pybind11.hpp:598]: This is a typo from when I wrote this caster.
Should be ``dpctl.program.SyclProgram``.

**Review by ndgrigorian — APPROVED** [review]
LGTM, thanks for writing the tests

File: dpctl/apis/include/dpctl4pybind11.hpp [hunk 2]
```diff
@@ -174,6 +192,8 @@ struct dpctl_capi
     std::shared_ptr<py::object> default_usm_memory;
     std::shared_ptr<py::object> default_usm_ndarray;
     std::shared_ptr<py::object> as_usm_memory;
+    std::shared_ptr<py::object> default_sycl_kernel;
```
Comment (diptorupd) [inline dpctl/apis/include/dpctl4pybind11.hpp:195]: Why are these not initialized like the other private members in the initializer list?

Comment (diptorupd) [reply dpctl/apis/include/dpctl4pybind11.hpp:195]: A nitpick. In general, `struct` should be restricted to passive objects that carry data and the `dpctl_capi` should be a `class`. Then the private members should follow a naming convention such as `std::shared_ptr<py::object> default_sycl_program_` with the trailing underscore. (Ref: https://google.github.io/styleguide/cppguide.html#Variable_Names)

File: dpctl/apis/include/dpctl4pybind11.hpp [hunk 3]
```diff
@@ -377,7 +405,7 @@ template <> struct type_caster<sycl::queue>
     bool load(handle src, bool)
     {
         PyObject *source = src.ptr();
-        auto &api = ::dpctl::detail::dpctl_capi::get();
+        auto const &api = ::dpctl::detail::dpctl_capi::get();
```
Comment (diptorupd) [inline dpctl/apis/include/dpctl4pybind11.hpp:408]: 👍

File: examples/pybind11/use_dpctl_sycl_kernel/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Usage of dpctl Entities in Pybind11
+
+## Description
+
+This extension demonstrates how you can use dpctl Python types,
+such as ``dpctl.SyclQueue``, in Pybind11
```
Comment (diptorupd) [inline examples/pybind11/use_dpctl_sycl_kernel/README.md:6]: typo?

File: examples/pybind11/use_dpctl_sycl_kernel/example.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# coding: utf-8
+
+import use_kernel as eg
+
+import dpctl
+import dpctl.program as dppr
+import dpctl.tensor as dpt
+
+q = dpctl.SyclQueue()
+
```
Comment (diptorupd) [inline examples/pybind11/use_dpctl_sycl_kernel/example.py:26]: Some documentation is warranted. I know what the example is doing, but not every user will know or should be expected to know just by reading the code.

**Review by diptorupd — APPROVED** [review]
LGTM.

I have a general comment about the state of `dpctl4pybind11.hpp`. The file needs API docs and we should generate the API doc and add it to our documentation page.

**Comment by oleksandr-pavlyk** [comment]
I am going to push suggested changes, which will dismiss approvals. I am going to wait for @chengjunlu's feedback and merge this tomorrow.

File: dpctl/apis/include/dpctl4pybind11.hpp (hunk 2 above)
Comment (oleksandr-pavlyk) [reply dpctl/apis/include/dpctl4pybind11.hpp]: Thanks @diptorupd. I have applied suggested changes. Incidentally `default_sycl_kernel` was unused, so I removed it. Thanks for spotting it.

File: examples/pybind11/use_dpctl_sycl_kernel/README.md (hunk 4 above)
Comment (oleksandr-pavlyk) [reply examples/pybind11/use_dpctl_sycl_kernel/README.md]: Fixed. Also output updated.

**Pull request merged**